src/balancer/src/codec.rs

```diff
@@ -0,0 +1,314 @@
+    }
+}
+
+trait Pgbuf: BufMut {
+    fn put_string(&mut self, s: &str);
+    fn put_length_i16(&mut self, len: usize) -> Result<(), io::Error>;
+    fn put_format_i8(&mut self, format: mz_pgwire_common::Format);
+    fn put_format_i16(&mut self, format: mz_pgwire_common::Format);
+}
+
+impl<B: BufMut> Pgbuf for B {
+    fn put_string(&mut self, s: &str) {
+        self.put(s.as_bytes());
+        self.put_u8(b'\0');
+    }
+
+    fn put_length_i16(&mut self, len: usize) -> Result<(), io::Error> {
+        let len = i16::try_from(len)
+            .map_err(|_| io::Error::new(io::ErrorKind::Other, "length does not fit in an i16"))?;
+        self.put_i16(len);
+        Ok(())
+    }
+
+    fn put_format_i8(&mut self, format: mz_pgwire_common::Format) {
+        self.put_i8(match format {
+            mz_pgwire_common::Format::Text => 0,
+            mz_pgwire_common::Format::Binary => 1,
+        })
+    }
+
+    fn put_format_i16(&mut self, format: mz_pgwire_common::Format) {
+        self.put_i8(0);
+        self.put_format_i8(format);
+    }
+}
+
+impl Decoder for Codec {
+    type Item = FrontendMessage;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<FrontendMessage>, io::Error> {
+        if src.len() > MAX_REQUEST_SIZE {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                format!(
+                    "request larger than {}",
+                    ByteSize::b(u64::cast_from(MAX_REQUEST_SIZE))
+                ),
+            ));
+        }
+        loop {
+            match self.decode_state {
+                DecodeState::Head => {
+                    if src.len() < 5 {
+                        return Ok(None);
+                    }
+                    let msg_type = src[0];
+                    let frame_len = parse_frame_len(&src[1..])?;
+                    src.advance(5);
+                    src.reserve(frame_len);
+                    self.decode_state = DecodeState::Data(msg_type, frame_len);
+                }
+
+                DecodeState::Data(msg_type, frame_len) => {
+                    if src.len() < frame_len {
+                        return Ok(None);
+                    }
+                    let buf = src.split_to(frame_len).freeze();
+                    let buf = Cursor::new(&buf);
+                    let msg = match msg_type {
+                        // Termination.
+                        b'X' => decode_terminate(buf)?,
+
+                        // Authentication.
+                        b'p' => decode_password(buf)?,
+
+                        // Invalid.
+                        _ => {
+                            return Err(io::Error::new(
+                                io::ErrorKind::InvalidData,
+                                format!("unknown message type {}", msg_type),
+                            ));
+                        }
+                    };
+                    src.reserve(5);
```

Comment:
5 seems to be a bit of a magic number here, presumably because the pgwire protocol has a format something like `[ msg_type <1>, <frame_len> 4, ... ]`? It's unclear to me why we reserve 5 bytes here though? Can you add a comment?

Reply:
This is copied directly from https://github.com/MaterializeInc/materialize/blob/a4b9a5220c5d957461b5a381ea85e0750edf80b9/src/pgwire/src/codec.rs#L486. I'd prefer to keep this code identical to make merging them in the future easier.